refactor(topology/metric_space/contracting): redefine using emetric

urkud opened the pull request:
TO CONTRIBUTORS:

Make sure you have:

  * [X] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [X] reviewed and applied [the documentation requirements](https://github.com/leanprover/mathlib/blob/master/docs/contribute/doc.md)
  * [ ] for tactics:
     * [ ] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [ ] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/test/tactics.lean)
  * [X] make sure definitions and lemmas are put in the right files
  * [X] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

urkud committed:
refactor(topology/metric_space/contracting): redefine using emetric

urkud committed:
Fix a typo produced by "copy+paste"

sgouezel commented on src/topology/metric_space/contracting.lean:
```diff
@@ -39,17 +46,164 @@ def contracting_with [emetric_space α] (K : ℝ≥0) (f : α → α) :=
+
+/-- Banach fixed-point theorem, contraction mapping theorem, `emetric_space` version.
+This version assumes that `f` is contracting in the whole space and sends a point `x`
+of a forward-invariant complete set `s` to the same ball of radius infinity.
```
This seems to me to be a weird set of assumptions: the most usual situation is when `f` is contracting on the whole space, and the most general one is when `f` is contracting on a forward invariant set. You are somewhat inbetween, assuming contraction on the whole space but arguing only in a forward invariant set. Do you have a motivating case for this intermediate degree of generality?

urkud replied:
I had a motivation but now I don't think it was valid. I need this theorem for a function contracting on a non-complete space covered by forward-invariant complete sets. In particular, I want to have uniqueness in this case.
I think that I'll prove this theorem (and define `efixed_point`) for a function contracting on a forward-invariant subset, then prove a version of uniqueness lemma for two forward-invariant sets.

P.S.: I'd prefer to prove it in a complete space, then apply to `λ x : s, ⟨f x, hfs x.2⟩`, where `hfs : maps_to f s s`, but conversion to/from subtype is not automatic.

urkud replied:
Done.

urkud committed:
Merge branch 'master' into emetric-contracting

urkud committed:
Merge branch 'master' into emetric-contracting

urkud committed:
Merge branch 'master' into emetric-contracting

urkud committed:
Fix compile

urkud committed:
Refactor `efixed_point`, `efixed_point'`

* Prove a version of Banach fixed point theorem for a map contracting
  on a complete forward-invariant set.

* Separately prove "primed" lemmas.

I Tried to define `efixed_point'` in terms of `efixed_point` and
failed: every time I use it, it generates a goal `complete_space ↥s`.
So, I decided to deduce `exists_fixed_point'` from
`exists_fixed_point`, then use it in the proofs.

sgouezel commented on src/data/subtype.lean:
```diff
@@ -30,6 +30,8 @@ end subtype
 namespace subtype
 variables {α : Sort*} {β : Sort*} {γ : Sort*} {p : α → Prop}
 
+lemma val_eq_coe (x : subtype p) : x.val = x := rfl
```
Should this be a simp lemma?

sgouezel replied:
Probably not, in fact: many places use `x.val` as the canonical form.

mergify[bot] committed:
Merge branch 'master' into emetric-contracting

mergify[bot] committed:
Merge branch 'master' into emetric-contracting
